### PR TITLE
Support --test/--bin/--lib in cargo-miri

### PR DESCRIPTION
**Pull request by divergentdave**
This PR addresses a FIXME in cargo-miri, and filters the targets to be checked when any of the `--bin`, '--test`, or `--lib` flags are passed.

### CODE REVIEW

File: cargo-miri/bin.rs [hunk 1]
```diff
@@ -378,6 +378,69 @@ path = "lib.rs"
...
+) -> (CargoTargets, Vec<String>, Vec<String>) {
+    let mut lib_present = false;
+    let mut bin_targets = Vec::new();
+    let mut test_targets = Vec::new();
+    let mut additional_args = Vec::new();
+    while let Some(arg) = args.next() {
+        match arg {
+            arg if arg == "--" => break,
+            arg if arg == "--lib" => lib_present = true,
+            arg if arg == "--bin" => {
```
Comment (RalfJung) [inline cargo-miri/bin.rs:411]: This looks like it supports `--bin name` but not `--bin=name`.

File: cargo-miri/bin.rs [hunk 2]
```diff
@@ -430,10 +496,7 @@ fn in_cargo_miri() {
             _ => continue,
         }
         // Forward user-defined `cargo` args until first `--`.
```
Comment (RalfJung) [inline cargo-miri/bin.rs:498]:
```suggestion
        // Forward further `cargo` args.
```

File: cargo-miri/bin.rs [hunk 3]
```diff
@@ -378,6 +378,69 @@ path = "lib.rs"
...
+fn parse_cargo_miri_args(
+    mut args: impl Iterator<Item = String>,
+) -> (CargoTargets, Vec<String>, Vec<String>) {
+    let mut lib_present = false;
+    let mut bin_targets = Vec::new();
+    let mut test_targets = Vec::new();
+    let mut additional_args = Vec::new();
+    while let Some(arg) = args.next() {
+        match arg {
+            arg if arg == "--" => break,
```
Comment (RalfJung) [inline cargo-miri/bin.rs:409]: Please add a comment saying that this is where the Miri arguments start.

**Comment by RalfJung** [comment]
Thanks a lot! This looks much cleaner than I thought was possible. ;)

However, this now applies to both `cargo miri run` and `cargo miri test`. Does that make sense? `cargo run` does not have `--lib` or `--test`, so maybe it would make more sense to exclude those. Also I think `cargo run` only allow `--bin` to be specified once.

On the other hand, I hope to get around to resolving https://github.com/rust-lang/miri/issues/739 within the next two weeks. So it does not seem worth the effort to spend a lot of time re-implementing cargo's logic here, if we are able to just reuse the existing logic.

File: cargo-miri/bin.rs [hunk 4]
```diff
@@ -398,22 +461,25 @@ fn in_cargo_miri() {
...
+            continue;
+        }
         // Now we run `cargo check $FLAGS $ARGS`, giving the user the
         // change to add additional arguments. `FLAGS` is set to identify
         // this target.  The user gets to control what gets actually passed to Miri.
         let mut cmd = cargo();
         cmd.arg("check");
         match (subcommand, kind.as_str()) {
             (MiriCommand::Run, "bin") => {
-                // FIXME: we just run all the binaries here.
```
Comment (RalfJung) [inline cargo-miri/bin.rs:415]: The first line here is still true -- `cargo miri run` runs all binaries when there are several, where `cargo run` picks `default-run` or errors of that key is not specified in `Cargo.toml`. So please don't fully remove this FIXME.

**Comment by divergentdave** [comment]
OK, made those changes. Reusing cargo's logic sounds great.

File: cargo-miri/bin.rs [hunk 5]
```diff
@@ -378,6 +378,74 @@ path = "lib.rs"
...
+                    if binary == "--" {
+                        show_error(format!("\"--bin\" takes one argument."));
+                    } else {
+                        bin_targets.push(binary)
+                    }
+                } else {
+                    show_error(format!("\"--bin\" takes one argument."));
+                }
+            }
+            arg if arg.starts_with("--bin=") => bin_targets.push((&arg[6..]).to_string()),
```
Comment (RalfJung) [inline cargo-miri/bin.rs:425]:
```suggestion
            arg if arg.starts_with("--bin=") => bin_targets.push((&arg["--bin".len()..]).to_string()),
```
Let's avoid magic constants. Same goes for `--test`.

File: cargo-miri/bin.rs [hunk 6]
```diff
@@ -398,22 +466,26 @@ fn in_cargo_miri() {
         return;
     }
 
+    let (target_filters, cargo_args, miri_args) =
```
Comment (RalfJung) [inline cargo-miri/bin.rs:469]:
```suggestion
    // FIXME: this accepts --test, --lib and multiple --bin for `cargo miri run`.
    let (target_filters, cargo_args, miri_args) =
```
I think it's okay to accept this for now, but we should leave a comment reminding us of the problem.

**Comment by RalfJung** [comment]
Thanks, this is great! Let's land this. @bors r+

It would also be good to have some testcases (cargo-miri is being tested by `test-cargo-miri/run-test.py`). I'll certainly add some when I try to adjust this to reuse the cargo logic, but if you want to help then already having some tests for the part that is covered by this PR would be great. :)

**Pull request merged**